File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last_seen = {} # Keep track of seen chars
        max_length = 0
        start = 0 # Start of window
        for end_ch in range(len(s)): # Move end of window
            if s[end_ch] in last_seen and start <= last_seen[s[end_ch]]: # Shift the start
                start = last_seen[s[end_ch]] + 1
            else: # Update the frequency list
                max_length = max(max_length, end_ch - start + 1)
            last_seen[s[end_ch]] = end_ch
        return max_length

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution 
{
    public int lengthOfLongestSubstring(String s) 
    {
        int longest = 0;
        if (s.length() < 1) // No string exists
            return longest;
        
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        for (int hi = 0, lo = 0; hi < s.length(); hi++)
        {
            if (map.containsKey(s.charAt(hi))) // Update start of substring
                lo = Math.max(lo, map.get(s.charAt(hi))+1);
            map.put(s.charAt(hi), hi); 
            longest = Math.max(longest, hi-lo+1); // Update max
        }
        return longest;
    }
}